substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,548 @@
+                false);
+
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    /* the map in which we place the result of a inlining decision */
+    private static final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Integer, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    /**
+     * Determines whether a call to a given method is to be inlined.
+     *
+     * @param b the context
+     * @param method the target method of an invoke
+     * @param args the arguments to the invoke
+     */
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+        if (b.parsingIntrinsic()) {
+            /* We are not interfering with any intrinsic method handling. */
+            return null;
+        }
+
+        if (method.getAnnotation(NeverInline.class) != null || method.getAnnotation(NeverInlineTrivial.class) != null) {
+            return null;
+        }
+
+        if (method.getAnnotation(RestrictHeapAccess.class) != null || method.getAnnotation(Uninterruptible.class) != null ||
+                b.getMethod().getAnnotation(RestrictHeapAccess.class) != null || b.getMethod().getAnnotation(Uninterruptible.class) != null) {
+            /*
+             * Caller or callee have an annotation that might prevent inlining. We don't check the
+             * exact condition but instead always bail out for simplicity.
+             */
+            return null;
+        }
+
+        CallSite callSite = new CallSite(toAnalysisMethod(b.getMethod()), b.bci());
+
+        InvocationResult inline;
+        if (analysis) {
+            BigBang bb = ((AnalysisBytecodeParser) b).bb;
+            TrivialMethodDetector detector = new TrivialMethodDetector(bb, providers, ((SharedBytecodeParser) b).getGraphBuilderConfig(), b.getOptions(), b.getDebug());
+            InvocationResult newResult = detector.analyzeMethod(callSite, (AnalysisMethod) method, args, null);
+            putIfAbsent(method, b.bci(), newResult);
+            inline = newResult;
+
+        } else {
+            inline = getResult(method, b.bci());
+        }
+
+        if (inline instanceof InvocationResultInline) {
+            InvocationResultInline inlineData = (InvocationResultInline) inline;
+            VMError.guarantee(inlineData.callee.equals(toAnalysisMethod(method)));
+
+            VMError.guarantee(((SharedBytecodeParser) b).inlineDuringParsingState == null);
+            ((SharedBytecodeParser) b).inlineDuringParsingState = inlineData;
+
+            if (analysis) {
+                AnalysisMethod aMethod = (AnalysisMethod) method;
+                aMethod.registerAsImplementationInvoked(null);
```

Comment:
Is it if we inlined the method?

Reply:
Why is this needed?

Reply:
If I remove it, there is a parsing error for non-recursive calls in the intrinsic org.graalvm.compiler.replacements.JDK9StringSubstitutions.equals(String, Object).

Reply:
What is the error? I would dig deeper on this.

Reply:
org.graalvm.compiler.debug.GraalError: All non-recursive calls in the intrinsic org.graalvm.compiler.replacements.JDK9StringSubstitutions.equals(String, Object) must be inlined or intrinsified: found call to String.coder()

Error is thrown by ReplacementsImpl for the method String.coder() if we don't call registerAsImplementationInvoked when we decide to inline this method.

Reply:
We are not interfering with any intrinsic method, while ReplacementsImpl forces inlining of intrinsics.  We decide to inline simple methods like String.coder() that has only constants nodes and return node in the graph and this method is also called by String.length() that is called by intrinsic JDK9StringSubstitutions.equals().